README
* source tree
    libmoss/
    |- include/
    |  |- moss/
    |    |- util.h
    |
    |- src/
    |  |- main.c
    |
    |- AUTHORS, ChangeLog, NEWS, README
    |- configure.ac, Makefile.am, autogen.sh, autodistclean.sh 

* libex1/include/ex1/util.h
    #ifndef _H_EX1_UTIL
    #define _H_EX1_UTIL
    
    #include <stdio.h>
    #include <stdlib.h>
    #include <unistd.h>
    #include <string.h>
    #include <stdarg.h>
    
    int ex1_test(const char *str);
    
    #ifdef __cplusplus
    } /* extern "C" */
    #endif
    
    #endif /* _H_EX1_UTIL */
    
* libex1/src/main.c
    #include <ex1/util.h>
    
    int ex1_test(const char *str)
    {
    	return printf("%s", str);
    }

* src/main.c
    #include <ex1/util.h>
    
    int main(int argc, char **argv)
    {
    	int i;
    
    	for (i = 0; i < argc; i++) {
    		log_debug("argv[%d / %d]: %s\n", i + 1, argc, argv[i]);
    	}
    
    	ex1_test("done\n");
    	return 0;
    }
* manual build/run
    $ gcc -o ex1 -I`pwd`/libex1/include src/main.c libex1/src/main.c
    
    $ ./ex1
    done

* initial configure.ac
    $ autoscan
    
    ex1/
    |- autoscan.log
    |- configure.scan

    $ rm autoscan.log # we ignore this file
    $ mv configure.scan configure.ac # template file
 
* configure.scan
    AC_PREREQ([2.69])
    AC_INIT([FULL-PACKAGE-NAME], [VERSION], [BUG-REPORT-ADDRESS])
    AC_CONFIG_SRCDIR([src/main.c])
    AC_CONFIG_HEADERS([config.h])
    
    AC_PROG_CC
    
    AC_CHECK_HEADERS([stdlib.h string.h unistd.h])
    
    AC_TYPE_SIZE_T
    
    AC_OUTPUT

* configure.ac without automake
    AC_PREREQ([2.69])
    AC_INIT([ex1], [1.1], [joelai@lavender])
    AC_CONFIG_SRCDIR([src/main.c])
    AC_CONFIG_HEADERS([config.h])
    
    AC_PROG_CC
    AC_PROG_LIBTOOL
    
    AC_CONFIG_FILES([Makefile])
    
    AC_TYPE_SIZE_T
   
    AC_OUTPUT

* test with autoconf
    autoreconf -fiv

* finally support autoconf/automake/libtool and configuration options


ChangeLog
2014-08-27  joelai  <joelai@lavender>

	* 
	  rename to libmoss
	* AUTHORS: New file.
	* autodistclean.sh: New file.
	* autogen.sh: New file.
	* configure.ac: New file.
	* include/Makefile.am: New file.
	* include/moss/Makefile.am: New file.
	* include/moss/util.h: New file.
	* Makefile.am: New file.
	* NEWS: New file.
	* README: New file.
	* src/main.c: New file.
	* src/Makefile.am: New file.

2014-08-13  joelai  <joelai@lavender>

	* configure.ac:
		generated by autoscan
		
	* **Makefile.mk: New file.
	* AUTHORS: New file.
	* NEWS: New file.
	* ChangeLog:
		make autoreconf -fiv happy

2014-08-13  joelai  <joelai@lavender>

	* libex1/include/ex1/util.h: New file.
	* libex1/src/main.c: New file.
	* src/main.c: New file.
	* README: 
		document before autotools
	* ChangeLog:
		format with Eclipse ChangeLog plugin


include/moss/util.h
#ifndef _H_MOSS_UTIL
#define _H_MOSS_UTIL

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <stdarg.h>

#define log_out(level, ...) do { \
	moss_log_out(level, __func__, __LINE__, __VA_ARGS__); \
} while(0)
#define log_error(...) log_out(1, __VA_ARGS__)
#define log_info(...) log_out(2, __VA_ARGS__)
#define log_debug(...) log_out(3, __VA_ARGS__)
#define log_verbose(...) log_out(4, __VA_ARGS__)

#define MOSS_STRINGIFY(_s) # _s
#define MOSS_STRINGIFY2(_s) MOSS_STRINGIFY(_s)

#ifdef __cplusplus
extern "C" {
#endif

int moss_vprintf(const char *fmt, va_list ap) __attribute__((format(printf, 1, 0)));

int moss_printf(const char *fmt, ...) __attribute__((format(printf, 1, 2)));
int moss_vsnprintf(char *o, size_t n,
		const char *fmt, va_list ap) __attribute__((format(printf, 3, 0)));
int moss_snprintf(char *o, size_t n,
		const char *fmt, ...) __attribute__((format(printf, 3, 4)));
int moss_log_out(int level, const char *func, long line,
		const char *fmt, ...) __attribute__((format(printf, 4, 5)));

/** @brief show memory in HEX/ASCII
 *
 * 00007FFF 00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F | 0123456789abcdef
 *          ----------------------------------------------- | ----------------
 * FFFFDDBE                                           1f 20 |               .
 * FFFFDDC0 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f 30 | !"#$%&'()*+,-./0
 */
int moss_showhex(const unsigned char *s, size_t n, unsigned long a);

#ifdef __cplusplus
} /* extern "C" */
#endif

#endif /* _H_MOSS_UTIL */


src/main.c
#include <moss/util.h>

int moss_vprintf(const char *fmt, va_list ap)
{
	int r;

	if (!fmt) return 0;

	r = vprintf(fmt, ap);
	if (r < 0) r = 0;
	return r;
}

int moss_printf(const char *fmt, ...)
{
	int r;
	va_list ap;

	va_start(ap, fmt);
	r = moss_vprintf(fmt, ap);
	va_end(ap);
	return r;
}

int moss_vsnprintf(char *o, size_t n, const char *fmt, va_list ap)
{
	int r;

	if (!o || n < 1) return 0;

	r = vsnprintf(o, n, fmt, ap);
	if (r < 0) r = 0;
	else if (r >= n) r = n - 1;
	o[r] = '\0';
	return r;
}

int moss_snprintf(char *o, size_t n, const char *fmt, ...)
{
	int r;
	va_list ap;

	va_start(ap, fmt);
	r = moss_vsnprintf(o, n, fmt, ap);
	va_end(ap);
	return r;
}

int moss_log_out(int level, const char *func, long line,
		const char *fmt, ...)
{
	const int len = 1024;
	int idx = 0;
	char msg[len];
	va_list ap;

	if (level == 1) {
		idx += moss_snprintf(msg + idx, len - idx, "ERROR %s #%ld ", func, line);
	} else if (level == 2) {
		idx += moss_snprintf(msg + idx, len - idx, "INFO %s #%ld ", func, line);
	} else if (level == 4) {
		idx += moss_snprintf(msg + idx, len - idx, "verbose %s #%ld ", func, line);
	} else {
		idx += moss_snprintf(msg + idx, len - idx, "Debug %s #%ld ", func, line);
	}
	if (idx >= len) goto finally;

	va_start(ap, fmt);
	idx += moss_vsnprintf(msg + idx, len - idx, fmt, ap);
	va_end(ap);
	if (idx >= len) goto finally;
finally:
	moss_printf("%s", msg);
	return idx;
}

int moss_showhex(const unsigned char *s, size_t n, unsigned long a)
{
#define SHOWHEX_ASCII 1
#define SHOWHEX_ADDR_WIDTH 8 /* sizeof(char*) * 2 */
#define SHOWHEX_ADDR_FMT "%08lX"
#define SHOWHEX_ADDR_SP "        "

/*                   1         2         3         4                   1
 * 12345678 12345678901234567890123456789012345678901234567   1234567890123456
 *
 * 00007FFF 00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F | 0123456789abcdef
 *          ----------------------------------------------- | ----------------
 * FFFFDDBE                                           1f 20 |               .
 * FFFFDDC0 21 22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f 30 | !"#$%&'()*+,-./0
 */
	char ascii_sp[16]  = "----------------";
	int i, l;

	moss_printf(SHOWHEX_ADDR_FMT " 00 01 02 03 04 05 06 07 08 09"
			" 0A 0B 0C 0D 0E 0F", a >> 32);
	a = (a << 32) >> 32;

	if (SHOWHEX_ASCII) moss_printf(" | 0123456789ABCDEF");
	moss_printf("\n" SHOWHEX_ADDR_SP " -----------------------------"
			"------------------");

	for (i = l = 0; i < n; l++) {
		if ((l & 0x0f) == 0) {
			ascii_sp[0x10] = '\0';
			if (SHOWHEX_ASCII) moss_printf(" | %s", ascii_sp);
			moss_printf("\n" SHOWHEX_ADDR_FMT, (unsigned long)(a + i));
		}
		if ((l & 0x0f) != ((unsigned long)(a + i) & 0x0f)) {
			moss_printf("   ");
			if (SHOWHEX_ASCII) ascii_sp[l % 0x10] = ' ';
		} else {
			moss_printf(" %02x", s[i]);
			if (SHOWHEX_ASCII) {
				ascii_sp[l % 0x10] = (isprint(s[i]) ? s[i] : '.');
			}
			i++;
		}
	}

	if (SHOWHEX_ASCII && ((l & 0x0f) != 0)) {
		for (; ; l++) {
			if ((l & 0x0f) == 0) {
				ascii_sp[l % 0x11] = '\0';
				moss_printf(" | %s\n", ascii_sp);
				break;
			}
			moss_printf("   ");
			ascii_sp[l % 0x10] = ' ';
		}
	}

	moss_printf("\n");

	return 0;
}
